src/index/Index.cpp

```diff
@@ -619,66 +683,70 @@ void Index::createPatterns(const string& fileName, const ExtVec& vec,
             << " would have to large a pattern." << std::endl;
+
   LOG(DEBUG) << "Total number of entities: "
              << (numEntitiesWithoutPatterns + numEntitiesWithPatterns)
              << std::endl;
 
-  // ensure neither of the relations is empty
-  if (entityHasPattern.size() == 0) {
-    entityHasPattern.push_back(std::array<Id, 2>{ID_NO_VALUE, ID_NO_VALUE});
-  }
-  if (entityHasRelation.size() == 0) {
-    entityHasRelation.push_back(std::array<Id, 2>{ID_NO_VALUE, ID_NO_VALUE});
-  }
+  LOG(DEBUG) << "Full has relation size: " << fullHasRelationSize << std::endl;
+  LOG(DEBUG) << "Full has relation entity multiplicity: "
+             << fullHasRelationMultiplicityEntities << std::endl;
+  LOG(DEBUG) << "Full has relation predicate multiplicity: "
+             << fullHasRelationMultiplicityPredicates << std::endl;
 
   // Store all data in the file
   ad_utility::File file(fileName.c_str(), "w");
-  off_t afterRelations = 0;
-  off_t afterPatterns = 0;
-  // write the has-pattern and has-relation relations into the file
-  auto md1 = writeRel(file, afterRelations, Id(0), entityHasPattern, true);
-  meta.add(md1.first, md1.second);
-  afterRelations = meta.getOffsetAfter();
-  auto md2 = writeRel(file, afterRelations, Id(1), entityHasRelation, false);
-  meta.add(md2.first, md2.second);
-  afterRelations = meta.getOffsetAfter();
-
-  size_t patternsPartSize = patterns.write(file);
-  afterPatterns = afterRelations + patternsPartSize;
-
-  // write the meta data into the file
-  file << meta;
-
-  // the last part of the file contains offset information for the varius parts
-  file.write(&afterRelations, sizeof(off_t));
-  file.write(&afterPatterns, sizeof(off_t));
+
+  // Write a byte of ones to make it less likely that an unversioned file is
+  // read as a versioned one (unversioned files begin with the id of the lowest
+  // entity that has a pattern).
+  // Then write the version, both multiplicitires and the full size.
+  unsigned char firstByte = 255;
+  file.write(&firstByte, sizeof(char));
+  file.write(&PATTERNS_FILE_VERSION, sizeof(uint32_t));
+  file.write(&fullHasRelationMultiplicityEntities, sizeof(double));
+  file.write(&fullHasRelationMultiplicityPredicates, sizeof(double));
+  file.write(&fullHasRelationSize, sizeof(size_t));
+
+  // write the entityHasPatterns vector
+  size_t numHasPatterns = entityHasPattern.size();
+  file.write(&numHasPatterns, sizeof(size_t));
+  file.write(entityHasPattern.data(), sizeof(Id) * numHasPatterns * 2);
+
+  // write the entityHasRelation vector
+  size_t numHasRelation = entityHasRelation.size();
+  file.write(&numHasRelation, sizeof(size_t));
+  file.write(entityHasRelation.data(), sizeof(Id) * numHasRelation * 2);
+
+  // write the patterns
+  patterns.write(file);
   file.close();
 
-  hasPattern.resize(entityHasPattern.back()[0] + 1);
-  // hasRelation.resize(entityHasRelation.back()[0] + 1);
+  LOG(INFO) << "Done creating patterns file." << std::endl;
```

Comment:
Maybe also print how many patterns we found

Reply:
I added the number of patterns to the rest of the info a couple of lines above this one.